src/extension/src/ActionHandler.py

```diff
@@ -64,32 +65,49 @@ def determine_operation(self, command):
             raise e
 
     def setup(self, action, log_message):
-        self.setup_file_logger(action)
-        self.setup_telemetry()
-        self.logger.log(log_message)
-        if action == Constants.ENABLE:
-            self.write_basic_status()
+        try:
+            self.setup_file_logger(action)
+            self.validate_sudo()  # Validating sudo as early as possible and before it's use in setup_telemetry
+            self.setup_telemetry()
+            self.logger.log(log_message)
+        except Exception:
+            raise
+        finally:
+            self.write_basic_status(action)
```

Comment:
curiosity: If lines 69 through 72 throw, can the exception be written to the basic status file and then rethrown?

It looks like you'll have to write basic status on line 73 at the end of the try and within the except. In that case a finally will no longer be necessary.

Reply:
With basic status, we set the top level status to "transitioning". Now incase of exceptions, and all these are currently terminal, the status is re-written to "error"

The way the code is in all setup() callers right now, we can have write_basic_status() on line 72, just before except, so a basic status (with status=Transitioning) is written in a happy path, and incase of exceptions, there is only 1 status write with status=error (instead of 2 writes). But thought of keeping write_basic_status() in finally, in case something changes in callers later